### ISSUE TITLE
Error while trying to do end-to-end EL 

### ISSUE THREAD

**Issue by ym-han** [informagi/REL#44]
Hi REL folks,

I've been trying to follow the end-to-end EL tutorial, but I'm running into the following error:

```
Traceback (most recent call last):
  File "rel_test.py", line 2, in <module>
    from REL.mention_detection import MentionDetection
  File "/users/yh31/data/yh31/envs/entity_linking/lib/python3.7/site-packages/REL/mention_detection.py", line 5, in <module>
    from REL.mention_detection_base import MentionDetectionBase
  File "/users/yh31/data/yh31/envs/entity_linking/lib/python3.7/site-packages/REL/mention_detection_base.py", line 5, in <module>
    from REL.utils import modify_uppercase_phrase, split_in_words
  File "/users/yh31/data/yh31/envs/entity_linking/lib/python3.7/site-packages/REL/utils.py", line 13, in <module>
    def fetch_model(path_or_url, cache_dir=flair.cache_root / "models/taggers"):
TypeError: unsupported operand type(s) for /: 'str' and 'str'
```

The code I tried running, which is basically just the tutorial code with my own paths, is:

```

from REL.mention_detection import MentionDetection
from REL.utils import process_results
from REL.entity_disambiguation import EntityDisambiguation
from REL.ner import Cmns, load_flair_ner

wiki_version = "wiki_2019"
base_url = "/users/yh31/data/yh31/packages/REL_data"


def example_preprocessing():
    # user does some stuff, which results in the format below.
    text = "Obama will visit Germany. And have a meeting with Merkel tomorrow."
    processed = {"test_doc1": [text, []], "test_doc2": [text, []]}
    return processed

input_text = example_preprocessing()


mention_detection = MentionDetection(base_url, wiki_version)
tagger_ner = load_flair_ner("ner-fast")
tagger_ngram = Cmns(base_url, wiki_version, n=5)
mentions_dataset, n_mentions = mention_detection.find_mentions(input_text, tagger_ner)


config = {
    "mode": "eval",
    "model_path": "/users/yh31/data/yh31/packages/REL_data/ed-wiki-2019",
}

model = EntityDisambiguation(base_url, wiki_version, config)
predictions, timing = model.predict(mentions_dataset)
print(predictions)
print(timing)

result = process_results(mentions_dataset, predictions, input_text)
print(result)
```

Am I making a mistake with the paths or something? Should I be using path objects instead of strings?

Apologies for this neophyte question! Looking forward to using your software!

**Comment by KDercksen** [comment]
Hi ym-han,

Thanks for your question! Interestingly, this is not something I can reproduce. Could you report back to me the output of:
```python
type(flair.cache_root)
```
As well as your version of Flair (`pip freeze | grep flair`)? `cache_root` should be a path object automatically, in which case the expression `flair.cache_root / "models/taggers"` is valid.

To force a fix, you might change the definition of `fetch_model` to `def fetch_model(path_or_url, cache_dir=Path(flair.cache_root) / "models/taggers")`

**Comment by ym-han** [comment]
Thanks, this is very helpful! You are right that there was an issue with flair and my installation --- after more testing I found that it had to do with how the virtual environment wasn't properly insulated because of the way I had set up my python paths. (I think there was another flair installation in a different virtual environment that was getting called.) But after getting the virtual environment properly insulated, the flair issue went away.

I did find, by the way, that I had to manually copy over models.json --- just doing the installation via the git clone method detailed on the main page hadn't copied that over, and the code needed that to run.

It would also be helpful for newbies if the docs could be more explicit about how "model_path" in `config` shouldn't be pointing to a subdirectory of the base url directory or anything like that. As far as I understand from experimentation, "model_path" is just supposed to be `ed-wiki-2019` (or 2014 if one wants to use that). But please let me know if I'm wrong!

Finally, it'd also be helpful for newbies if you could provide the prediction scores that you'd got for a bunch of sample data, just so we can check that we are getting the numbers we 'should' be getting --- just so we can be sure that we've installed everything correctly.

P.S.: While going through the repo, I saw that you were benchmarking REL against BLINK. Could I ask how REL fared? (My experience with BLINK was that the interactive thing worked, but the sample code they provided for programmatically / non-interactively supplying arguments to the entity linker seemed buggy.)

**Comment by KDercksen** [comment]
> Thanks, this is very helpful! You are right that there was an issue with flair and my installation --- after more testing I found that it had to do with how the virtual environment wasn't properly insulated because of the way I had set up my python paths. (I think there was another flair installation in a different virtual environment that was getting called.) But after getting the virtual environment properly insulated, the flair issue went away.

Good to hear that's fixed! :-)

> I did find, by the way, that I had to manually copy over models.json --- just doing the installation via the git clone method detailed on the main page hadn't copied that over, and the code needed that to run. But again I could have just set things up wrongly or something.

Hmm, that shouldn't be happening... It is possible that it's due to your installation. You could try to reinstall in a clean virtualenv and see if that fixes it automatically, but it sounds like you already solved it for now.

> It would also be helpful for newbies if the docs could be more explicit about how "model_path" in `config` shouldn't be pointing to a subdirectory of the base url directory or anything like that. As far as I understand from experimentation, "model_path" is just supposed to be `ed-wiki-2019` (or 2014 if one wants to use that). But please let me know if I'm wrong!

Thanks for the tip. We touch on this in the [tutorial on using custom models](https://github.com/informagi/REL/blob/master/tutorials/07_custom_models.md), but I can understand how it is not immediately obvious from the earlier tutorials. Using just `ed-wiki-2019` will download the ED model into your local cache, for example. If you download the model yourself, you should use a full path without extension; e.g. if your ED model resides in `/path/to/model_dir` with the files `/path/to/model_dir/model.config` and `/path/to/model_dir/model.state_dict`, you should pass `/path/to/model_dir/model` in the ED config dictionary. Hopefully that clears it up a little!

> Finally, it'd also be helpful for newbies if you could provide the prediction scores that you'd got for a bunch of sample data, just so we can check that we are getting the numbers we 'should' be getting --- just so we can be sure that we've installed everything correctly.

This would be a nice addition, thanks for the idea!

> P.S.: While going through the repo, I saw that you were benchmarking REL against BLINK. Could I ask how REL fared? (My experience with BLINK was that the interactive thing worked, but the sample code they provided for programmatically / non-interactively supplying arguments to the entity linker seemed buggy.)

As far as I remember, REL did a bit better/quicker; maybe @mickvanhulst can chime in, as he did the comparison.

Thanks for your suggestions, we're happy to see some community involvement in REL! Please feel free to ask for further help if you happen to need it. Any other suggestions are also welcome!

**Comment by mickvanhulst** [comment]
Hi @KDercksen and @ym-han,

I did indeed test BLINK briefly. I agree that the sample code was a bit buggy, but I did get it to work. In the end I was not able to evaluate BLINK on GERBIL as it seemed that they relied on accepting entire corpuses in one go, whereas the GERBIL API only provided the user with one document at a time. This made the API extremely slow and after trying a few things out, I decided to leave it at that.
